rcl/include/rcl/allocator.h

```diff
@@ -0,0 +1,62 @@
+// Copyright 2015 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL__ALLOCATOR_H_
+#define RCL__ALLOCATOR_H_
+
+#if __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/types.h"
+#include "rcl/visibility_control.h"
+
+/// Encapsulation of an allocator.
+/* To use malloc, free, and realloc use rcl_get_default_allocator(). */
+typedef struct rcl_allocator_t
+{
+  /// Allocate memory, given a size and state structure.
+  /* An error should be indicated by returning NULL. */
+  void * (*allocate)(size_t size, void * state);
+  /// Deallocate previously allocated memory, mimicking free().
+  void (* deallocate)(void * pointer, void * state);
+  /// Reallocate if possible, otherwise it deallocates and allocates.
+  /* If unsupported then do deallocate and then allocate.
+   * \TODO(wjwwood): should this behave as reallocf?
+   * This should behave as realloc is described, as opposed to reallocf, i.e.
+   * the memory given by pointer will not be free'd automatically if realloc
+   * fails.
```

Comment:
Well the `reallocf` behavior was introduced because the pattern for safely using `realloc` is a bit strange. It often leads to this subtle bug:

```c++
void * mem = malloc(...);
mem = realloc(mem, ...);
if (!mem) {
  // The memory allocated by malloc is now leaked,
  // since it was not free'd and the only pointer to it was overwritten with NULL.
}
```

Which leads to something correct being like this:

```c++
void * mem = malloc(...);
void * new_mem = realloc(mem, ...);
if (!new_mem) {
  free(mem);
}
mem = new_mem;
```

And `reallocf` avoids this issue completely with this pattern now being correct:

```c++
void * mem = malloc(...);
mem = reallocf(mem, ...);
if (!mem) {
  // No memory has been leaked.
}
```

The only downside to only having `reallocf` would be that you cannot keep the original memory if `realloc` fails, which might be a valid thing to do.

I think I'll instead make a wrapper function which does something similar to this:

```c++
void *
rcl_reallocf(
  void * (*realloc_func)(void *, size_t, void *),
  void (*free_func)(void *, void *),
  void * original_mem, size_t new_size, void * state)
{
  void * new_mem = realloc_func(original_mem, new_size, state);
  if (!new_mem) {
    free_func(original_mem, state);
  }
  return new_mem;
}
```